src/lib.rs

```diff
@@ -0,0 +1,902 @@
+    /// (which is sometimes useful for post-allocation analyses).
+    ///
+    /// kind:3 unused:1 index:28
+    bits: u32,
+}
+
+impl std::fmt::Debug for Allocation {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        std::fmt::Display::fmt(self, f)
+    }
+}
+
+impl std::fmt::Display for Allocation {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self.kind() {
+            AllocationKind::None => write!(f, "none"),
+            AllocationKind::Reg => write!(f, "{}", self.as_reg().unwrap()),
+            AllocationKind::Stack => write!(f, "{}", self.as_stack().unwrap()),
+        }
+    }
+}
+
+impl Allocation {
+    #[inline(always)]
+    pub(crate) fn new(kind: AllocationKind, index: usize) -> Self {
+        assert!(index < (1 << 28));
+        Self {
+            bits: ((kind as u8 as u32) << 29) | (index as u32),
+        }
+    }
+
+    #[inline(always)]
+    pub fn none() -> Allocation {
+        Allocation::new(AllocationKind::None, 0)
+    }
+
+    #[inline(always)]
+    pub fn reg(preg: PReg) -> Allocation {
+        Allocation::new(AllocationKind::Reg, preg.index())
+    }
+
+    #[inline(always)]
+    pub fn stack(slot: SpillSlot) -> Allocation {
+        Allocation::new(AllocationKind::Stack, slot.0 as usize)
+    }
+
+    #[inline(always)]
+    pub fn kind(self) -> AllocationKind {
+        match (self.bits >> 29) & 7 {
+            5 => AllocationKind::None,
+            6 => AllocationKind::Reg,
+            7 => AllocationKind::Stack,
+            _ => unreachable!(),
+        }
+    }
+
+    #[inline(always)]
+    pub fn is_none(self) -> bool {
+        self.kind() == AllocationKind::None
+    }
+
+    #[inline(always)]
+    pub fn is_reg(self) -> bool {
+        self.kind() == AllocationKind::Reg
+    }
+
+    #[inline(always)]
+    pub fn is_stack(self) -> bool {
+        self.kind() == AllocationKind::Stack
+    }
+
+    #[inline(always)]
+    pub fn index(self) -> usize {
+        (self.bits & ((1 << 28) - 1)) as usize
+    }
+
+    #[inline(always)]
+    pub fn as_reg(self) -> Option<PReg> {
+        if self.kind() == AllocationKind::Reg {
+            Some(PReg::from_index(self.index()))
+        } else {
+            None
+        }
+    }
+
+    #[inline(always)]
+    pub fn as_stack(self) -> Option<SpillSlot> {
+        if self.kind() == AllocationKind::Stack {
+            Some(SpillSlot(self.index() as u32))
+        } else {
+            None
+        }
+    }
+
+    #[inline(always)]
+    pub fn bits(self) -> u32 {
+        self.bits
+    }
+
+    #[inline(always)]
+    pub fn from_bits(bits: u32) -> Self {
+        Self { bits }
+    }
+}
+
+// N.B.: These values must be *disjoint* with the values used to
+// encode `OperandPolicy`, because they share a 3-bit field.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(u8)]
+pub enum AllocationKind {
+    None = 5,
```

Comment:
It isn't right now, but I can't fully convince myself there won't be a case in the future where it might be useful; e.g. maybe there will be some operand mode where it's acceptable to fail to find an allocation. I guess it's also useful to expose an "empty" default for initialization purposes.